src/Normalizer/ContentEntityNormalizer.php

```diff
@@ -263,190 +313,238 @@ public function denormalize($data, $class, $format = NULL, array $context = arra
+        $translations[$key] = $this->denormalizeTranslation($item);
       }
     }
 
+    // @todo {@link https://www.drupal.org/node/2599946 Move the below update
+    // logic to the resource plugin instead.}
+    $storage = $this->entityManager->getStorage($this->entity_type_id);
+
+
     // @todo {@link https://www.drupal.org/node/2599926 Use the passed $class to instantiate the entity.}
 
-    if ($entity_id) {
-      if ($entity = $storage->load($entity_id) ?: $storage->loadDeleted($entity_id)) {
-        foreach ($data as $name => $value) {
-          if ($name == 'default_langcode') {
-            continue;
+    if ($this->entity_id) {
+      if ($entity = $storage->load($this->entity_id) ?: $storage->loadDeleted($this->entity_id)) {
+        if (!empty($translations[$entity->language()->getId()])) {
+          foreach ($translations[$entity->language()->getId()] as $name => $value) {
+            if ($name == 'default_langcode') {
+              continue;
+            }
+            $entity->{$name} = $value;
           }
-          $entity->{$name} = $value;
         }
       }
-      elseif (isset($data[$id_key])) {
-        unset($data[$id_key], $data[$revision_key]);
+      elseif (isset($translations[$default_langcode][$this->id_key])) {
+        unset($translations[$default_langcode][$this->id_key], $translations[$default_langcode][$revision_key]);
         $entity_id = NULL;
-        $entity = $storage->create($data);
+        $entity = $storage->create($translations[$default_langcode]);
+      }
+
+      foreach ($site_languages as $site_language) {
+        $langcode = $site_language->getId();
+        if ($entity->language()->getId() != $langcode) {
+          $entity->addTranslation($langcode, $translations[$langcode]);
+        }
       }
     }
     else {
       $entity = NULL;
-      $entity_types_to_create = ['replication_log', 'user'];
-      if (!empty($bundle_key) && !empty($data[$bundle_key]) || in_array($entity_type_id, $entity_types_to_create)) {
-        unset($data[$id_key], $data[$revision_key]);
-        $entity = $storage->create($data);
+      $this->entity_types_to_create = ['user'];
+      if (!empty($this->bundle_key) && !empty($translations[$default_langcode][$this->bundle_key]) || in_array($this->entity_type_id, $this->entity_types_to_create)) {
+        unset($translations[$default_langcode][$this->id_key], $translations[$default_langcode][$revision_key]);
+        $entity = $storage->create($translations[$default_langcode]);
       }
     }
 
-    if ($entity_id) {
+    if ($this->entity_id) {
       $entity->enforceIsNew(FALSE);
       $entity->setNewRevision(FALSE);
       $entity->_rev->is_stub = FALSE;
     }
 
-    Cache::invalidateTags(array($entity_type_id . '_list'));
+    Cache::invalidateTags(array($this->entity_type_id . '_list'));
 
     return $entity;
   }
 
+  /**
+   * @param $translation
+   * @return mixed
+   */
+  private function denormalizeTranslation($translation) {
```

Comment:
Instead of using parameters on `$this` I would prefer that the method receive all parameters as arguments to the method. That way the method doesn't rely too much on the state of the object.